src/orion/client/runner.py

```diff
@@ -0,0 +1,323 @@
+        ]
+        return "\n".join(lines)
+
+
+class Runner:
+    """Run the optimization process given the current executor"""
+
+    def __init__(
+        self,
+        client,
+        fct,
+        pool_size,
+        idle_timeout,
+        max_trials_per_worker,
+        max_broken,
+        trial_arg,
+        on_error,
+        interrupt_signal_code=None,
+        **kwargs,
+    ):
+        self.client = client
+        self.fct = fct
+        self.pool_size = pool_size
+        self.max_trials_per_worker = max_trials_per_worker
+        self.max_broken = max_broken
+        self.trial_arg = trial_arg
+        self.on_error = on_error
+        self.kwargs = kwargs
+
+        self.gather_timeout = 0.01
+        self.idle_timeout = idle_timeout
+
+        self.worker_broken_trials = 0
+        self.trials = 0
+        self.futures = []
+        self.pending_trials = dict()
+        self.free_worker = pool_size
+        self.stat = _Stat()
+
+        if interrupt_signal_code is None:
+            interrupt_signal_code = orion.core.config.worker.interrupt_signal_code
+
+        self.interrupt_signal_code = interrupt_signal_code
+
+    @property
+    def is_done(self):
+        """Returns true if the experiment has finished."""
+        return self.client.is_done
+
+    @property
+    def has_remaining(self):
+        """Returns true if the worker can still pick up work"""
+        return (
+            self.max_trials_per_worker - (self.trials - self.worker_broken_trials) > 0
+        )
+
+    @property
+    def is_idle(self):
+        """Returns true if none of the workers are running a trial"""
+        return len(self.pending_trials) <= 0
+
+    @property
+    def running(self):
+        """Returns true if we are still running trials."""
+        return self.pending_trials or (self.has_remaining and not self.is_done)
+
+    def run(self):
+        """Run the optimizing process until completion.
+
+        Returns
+        -------
+        the total number of trials processed
+
+        """
+        idle_start = time.time()
+        idle_end = 0
+        idle_time = 0
+
+        while self.running:
+
+            # Get new trials for our free workers
+            with self.stat.time("sample"):
+                new_trials = self.sample()
+
+            # Scatter the new trials to our free workers
+            with self.stat.time("scatter"):
+                self.scatter(new_trials)
+
+            # Gather the results of the workers that have finished
+            with self.stat.time("gather"):
+                self.gather()
+
+            if self.is_idle:
+                idle_end = time.time()
+                idle_time += idle_end - idle_start
+                idle_start = idle_end
+
+                log.debug(f"Workers have been idle for {idle_time:.2f} s")
+            else:
+                idle_start = time.time()
+                idle_time = 0
+
+            if self.is_idle and idle_time > self.idle_timeout:
```

Comment:
Should add tests for idle_timeout for each use case supported below (if pending or has remaining).

Reply:
I added some tests for it